src/osd/scrubber/scrub_machine.h

```diff
@@ -314,12 +368,45 @@ struct NotActive : sc::state<NotActive, ScrubMachine>, NamedSimply {
   sc::result react(const AfterRepairScrub&);
 };
 
-struct ReservingReplicas : sc::state<ReservingReplicas, ScrubMachine>,
+
+/**
+ *  Session
+ *
+ *  This state encompasses the two main "active" states: ReservingReplicas and
+ *  ActiveScrubbing.
+ *  'Session' is the owner of all the resources that are allocated for a
+ *  scrub session.
+ *
+ *  Exit from this state is either following an interval change (with two
+ *  events expected - interval-changed & full-reset, or with
+ *  'FullReset' (that would cover all other completion/termination paths).
+ *  Note that if terminating the session following an interval change - no
+ *  reservations are released. This is because we know that the replicas are
+ *  also resetting their reservations.
+ */
+struct Session : sc::state<Session, ScrubMachine, ReservingReplicas>,
+                 NamedSimply {
+  explicit Session(my_context ctx);
+  ~Session();
+
+  using reactions = mpl::list<sc::transition<FullReset, NotActive>,
+                              sc::custom_reaction<IntervalChanged>>;
+
+  sc::result react(const IntervalChanged&);
+
+  /// managing the scrub session's reservations (optiona, as
+  /// it's an RAII wrapper around the state of 'holding reservations')
+  std::optional<ReplicaReservations> m_reservations{std::nullopt};
```

Comment:
First, an initially constructed ReplicaReservations should be represent "no replicas reserved thus far", which I think is what std::nullopt should represent as well.  No need for this to be an optional.

More significantly, I don't see any purpose for the ReplicaReservations class.  It requires a bunch of boilerplate code for capturing interfaces already held by the scrubber.  With the comments I made above, it should have only three real members:
- m_sorted_secondaries (rename to replicas_to_reserve)
- m_next_to_request (rename to next_replica_to_reserve)
- m_request_sent_at (rename to last_reservation_request_sent_at or something)

I think if you simply move the members and methods directly to Session, the code will be vastly easier to scan and no more complex as Session really doesn't do much else.